Update WAVE test runner

FritzHeiden opened the pull request:
This PR updates the wave test runner used in WMAS and DPCTF projects by CTA. It mainly includes bug fixes as well as change to support python 3.8

FritzHeiden committed:
feat: update wave test runner to current version

foolip commented:
@FritzHeiden is there someone familiar with this setup who you'd like review from? If not, then as long as the CI passes and the diff looks reasonable, someone like me could rubberstamp and merge it. Right now the CI isn't passing, have you checked that out?

JohnRiv commented:
@foolip as the WAVE HATF chair, I'll review this PR for the implementation. Agree we need tests to pass as well for this to be able to be merged.

FritzHeiden commented:
Unfortunately, I don't know how to resolve these errors:

```
[Errno 2] No such file or directory: '/home/test/web-platform-tests/tools/wave/../wptrunner/requirements_chrome.txt'
```

This is not part of the `tools/wave` directory, so I don't know what this is for and how to resolve it. Can someone please assist me with this?

JohnRiv commented on tools/wave/ecmascript/generate-tests.js:
```diff
@@ -0,0 +1,309 @@
+    }
+
+    const jsRelativePath = path.relative(testsPath, currentPath);
+    // console.log(jsRelativePath.replace('.js', ''))
```
probably best to remove commented-out debug code to clean up the files

JohnRiv commented on tools/wave/ecmascript/generate-tests.js:
```diff
@@ -0,0 +1,309 @@
+    const jsSrc = await readFile(currentPath);
+    const meta = parseFrontmatter(jsSrc);
+    const includes = (meta && meta.includes) || [];
+    //console.log(includes);
```
remove commented-out debug code

JohnRiv commented on tools/wave/ecmascript/test-template.html:
```diff
@@ -0,0 +1,13 @@
+<title>{{ TEST_TITLE }}</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<!-- <script charset="UTF-8" src="/ecmascript/harness/jquery-1.4.2.min.js"></script> -->
```
removed commented-out code (I'm not sure this path to jQuery is valid if it was uncommented?)

JohnRiv commented on tools/wave/ecmascript/test-template.iframe.html:
```diff
@@ -0,0 +1,16 @@
+<script charset="UTF-8" src="/ecmascript/harness/cth.js"></script>
+<script charset="UTF-8" src="/ecmascript/harness/propertyHelper.js"></script>
+<script charset="UTF-8" src="/ecmascript/harness/PromiseHelper.js"></script>
+<!-- <script charset="UTF-8" src="/ecmascript/harness/jquery-1.4.2.min.js"></script> -->
```
removed commented-out code

JohnRiv commented on tools/wave/ecmascript/webplatform-adapter.js:
```diff
@@ -0,0 +1,109 @@
+document.getElementById("iframe").contentWindow.addEventListener("error", errorEventListener);
+document.getElementById("iframe").contentWindow.$ERROR = $ERROR;
+
+// /ecmascript/tests/built-ins/RegExp/prototype/Symbol.match/builtin-coerce-global.html "Aw Snap"
+// /ecmascript/tests/built-ins/RegExp/prototype/Symbol.match/coerce-global.html "Aw Snap"
+// /ecmascript/tests/built-ins/RegExp/prototype/Symbol.replace/coerce-global.html "Aw Snap"
+// /ecmascript/tests/language/statements/for-of/iterator-next-reference.html "Website unresponsive"
```
unclear if we need these comments to stay in the code or not

JohnRiv commented on tools/wave/requirements.txt:
```diff
@@ -1,2 +1,2 @@
-ua-parser==0.18.0
-python-dateutil==2.9.0.post0
+ua-parser==0.8.0
+python-dateutil==2.8.1
```
was this change intentional? the previous versions [were set by dependabot](https://github.com/web-platform-tests/wpt/commits/master/tools/wave/requirements.txt), so I assume this will just cause dependabot to issue a follow-up PR

JohnRiv commented on tools/wave/tox.ini:
```diff
@@ -1,14 +1,13 @@
 [tox]
-envlist = py38,py39,py310,py311,py312
+envlist = py36,py37,py38,py39
```
[prior commits to this file](https://github.com/web-platform-tests/wpt/commits/master/tools/wave/tox.ini) from the past couple years indicate this `envlist` change should be reverted

JohnRiv commented on tools/wave/tox.ini:
```diff
@@ -1,14 +1,13 @@
-  -r{toxinidir}/requirements.txt
   -r{toxinidir}/../wptrunner/requirements.txt
-  -r{toxinidir}/../wptrunner/requirements_chromium.txt
+  -r{toxinidir}/../wptrunner/requirements_chrome.txt
```
I believe this should be reverted, the filename is [requirements_chromium.txt](https://github.com/web-platform-tests/wpt/blob/master/tools/wptrunner/requirements_chromium.txt)

JohnRiv reviewed (changes requested):
@FritzHeiden I left some comments inline. I believe the last few may resolve most/all of the failing tests

FritzHeiden committed:
style: remove comments

FritzHeiden committed:
chore: revert changes in requirements.txt and tox.ini

JohnRiv reviewed (changes requested):
There are a few other unit test failures besides `no-untyped-def`, but this should at least address the majority of them

FritzHeiden committed:
fix: mypy annotations and future imports

JohnRiv commented on tools/wave/network/api/api_handler.py:
```diff
@@ -5,13 +5,16 @@
-from urllib.parse import parse_qsl
+try:
+    from urllib.parse import parse_qsl
+except ImportError:
+    from urlparse import parse_qsl
```
Do you need this? It's throwing 2 errors:

> error: Cannot find implementation or library stub for module named "urlparse" [import-not-found]
> error: Name "parse_qsl" already defined (possibly by an import) [no-redef]

and other parts of the project use `urllib.parse` and [`parse_qsl` is in python 3.8](https://docs.python.org/3.8/library/urllib.parse.html#urllib.parse.parse_qsl), so I assume that's fine?

JohnRiv commented on tools/wave/network/api/tests_api_handler.py:
```diff
@@ -2,7 +2,10 @@
 import json
 
-from urllib.parse import urlunsplit
+try:
+    from urllib.parse import urlunsplit
+except ImportError:
+    from urlparse import urlunsplit
```
This is also throwing 2 errors:

> error: Cannot find implementation or library stub for module named "urlparse" [import-not-found]
> error: Name "urlunsplit" already defined (possibly by an import) [no-redef]

likewise, do you need the `except`?

JohnRiv reviewed (changes requested):
@FritzHeiden if you address the inline comments here, I believe that should take care of all the failures we see in tools/ unittests

FritzHeiden committed:
fix: wpt CI pipeline errors

FritzHeiden committed:
fix: flake8 errors

FritzHeiden committed:
fix: tests not working with tox

FritzHeiden committed:
fix: untyped calls mypy error in wave-cli.py

FritzHeiden committed:
fix: remove unused tests

FritzHeiden commented:
@JohnRiv Thanks for you help. I removed the two test files and the error still occurs. Do you have another idea how to prevent the pipeline from trying to run tests in the wave directory?

JohnRiv commented:
@FritzHeiden it appears this is what executes the command to run those tests:

https://github.com/web-platform-tests/wpt/blob/master/tools/ci/ci_tools_integration_test.sh#L16-L20

so that code will have to be removed as well (or the tests should be fixed)

FritzHeiden committed:
fix: remove wave test integration

FritzHeiden commented:
@JohnRiv Removing these lines resolves the issues from the CI. The other two failing checks don't seem to be related to the changes in this PR.

JohnRiv commented:
Thanks @FritzHeiden. The failures I see in the logs are ones I see in other PRs, so it definitely seems unrelated to the changes here:
```
web-platform-test
~~~~~~~~~~~~~~~~~
Ran 379 checks (214 subtests, 165 tests)
Expected results: 366
Skipped: 8 tests
Unexpected results: 5
  test: 5 (5 crash)

Unexpected Results
------------------
/infrastructure/assumptions/canvas-background.html
  CRASH /infrastructure/assumptions/canvas-background.html
/infrastructure/reftest/reftest.www.sub.html
  CRASH /infrastructure/reftest/reftest.www.sub.html
/infrastructure/reftest/reftest_ref_timeout.html
  CRASH /infrastructure/reftest/reftest_ref_timeout.html
/infrastructure/reftest/reftest_timeout.html
  CRASH /infrastructure/reftest/reftest_timeout.html
/infrastructure/crashtests/testdriver.html
  CRASH /infrastructure/crashtests/testdriver.html
```

JohnRiv reviewed (approved):
Changes look good from a CTA WAVE HATF perspective 👍

JohnRiv commented:
@foolip @jgraham @DanielRyanSmith, these changes are good from WAVE's perspective. 
Do you have any requested changes, and if not, can this be merged in?

JohnRiv commented:
Here is a copy of the full log from the failing **Run tests (Safari Technology Preview)** job in case it is no longer available when someone is available to review this PR: [20250320.18.191.SafariTechnologyPreview.log](https://github.com/user-attachments/files/19437609/20250320.18.191.SafariTechnologyPreview.log)

louaybassbouss commented:
@foolip @jgraham @DanielRyanSmith please let us know if anything you need from us (@FritzHeiden  and me).

JohnRiv commented:
@foolip @jgraham @DanielRyanSmith I wanted to check in on the PR review for this. Anything you need from the WAVE team?

gsnedders commented on tools/wave/data/client.py:
```diff
@@ -1,6 +1,6 @@
 # mypy: allow-untyped-defs
 
-class Client:
+class Client(object):
```
This is equivalent on Python 3.

gsnedders commented on tools/wave/data/session.py:
```diff
@@ -9,8 +9,11 @@
 
-class Session:
+
+class Session(object):
```
This is equivalent on Python 3.

gsnedders commented on tools/wave/network/api/api_handler.py:
```diff
@@ -68,29 +68,29 @@ def create_hal_list(self, items, uris, index, count, total):
             if "self" in uris:
                 self_uri = uris["self"]
-                self_uri += f"?index={index}&count={count}"
+                self_uri += "?index={}&count={}".format(index, count)
```
It's unclear to me why this is being changed away from f-strings? This is all supported from Python 3.6 onwards.

gsnedders commented on tools/wave/network/api/devices_api_handler.py:
```diff
@@ -91,7 +91,7 @@ def register_global_event_listener(self, request, response):
             query = self.parse_query_parameters(request)
 
-            if "device_token" in query:
+            if u"device_token" in query:
```
This is equivalent on Python 3.

gsnedders commented on tools/wave/network/api/devices_api_handler.py:
```diff
@@ -168,18 +168,18 @@ def handle_request(self, request, response):
                 self.create_device(request, response)
                 return
-            if method == "GET":
+            if method == u"GET":
```
This is equivalent on Python 3.

gsnedders reviewed (changes requested):
While I realise this is mostly just landed what already exists upstream, I'm kinda concerned by the direction this is taking — removing tests, adding a `.python-version` that claims Python 3.8 is required while at the same time seemingly re-introducing Python 2 support. I didn't go through all the files, but the overwhelming majority seem to just be repeating themselves again and again.

I'm reasonably confident the CI failure here is unrelated to this PR, however.

FritzHeiden committed:
refactor: remove object parameter from class definitions

FritzHeiden committed:
refactor: remove redundant io.open imports

FritzHeiden committed:
refactor: remove redundant super call paramters

FritzHeiden committed:
refactor: remove explicit u string declaration

FritzHeiden committed:
refactor: use f-strings

FritzHeiden committed:
refactor: use property instead of index in urlsplit

Co-authored-by: Sam Sneddon <me@gsnedders.com>

DanielRyanSmith commented:
@gsnedders Last call, do you want another review here?

gsnedders reviewed (changes requested):
Have a singular comment, but also: this is failing the CI jobs, and they need to be fixed.

FritzHeiden committed:
chore: reintroduce tools/wave/tox.ini into pipeline

FritzHeiden committed:
chore: reintroduce wave tests

FritzHeiden committed:
feat: use current version of test

FritzHeiden committed:
chore: add dateutil to tools/requirements_mypy.txt

gsnedders committed:
Update .python-version

Update this after https://github.com/web-platform-tests/wpt/commit/e974e2c9312f083dda61339ec5f45a0ef22a2f3f

gsnedders reviewed (approved):
Having fixed `.python-version` after we bumped that in https://github.com/web-platform-tests/wpt/commit/e974e2c9312f083dda61339ec5f45a0ef22a2f3f, I'll rubber-stamp this, without actually reviewing it. The changes outside of `tools/wave` look fine.
